### PR TITLE
Merge implemented tests for LoginServlet and UserFoldersServlet to main branch

### PR DESCRIPTION
**Pull request by ethyus**
Test results:
Screenshot 2020-08-10 at 3 51 08 PM

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 1]
```diff
@@ -24,6 +26,7 @@
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
 
     UserService userService = UserServiceFactory.getUserService();
+    Map<String, Object> jsonInfo = new HashMap<>();
```
Comment (ESaracay) [inline src/main/java/com/google/sps/servlets/LoginServlet.java]: Should move this down to where it's used and add a comment explaining why you're using a hashMap.

File: src/test/java/com/google/sps/UserFoldersServletTest.java [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+  @After
+  public void tearDown() throws Exception {
+    helper.tearDown();
+  }
+  
+  @Test
+  public void QueryUserFolders() throws Exception {
+    
+    // Generate testing User
```
Comment (ESaracay) [inline src/test/java/com/google/sps/UserFoldersServletTest.java]: I still feel that comments can be more specific, like what is UserA and the fields it has. The test explanation's seem very generic and I'm not sure whats being tested.

File: src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 3]
```diff
@@ -34,46 +37,52 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
     if (userService.isUserLoggedIn()) {
-      userEmail = userService.getCurrentUser().getEmail();
       userId = userService.getCurrentUser().getUserId();
+      userEmail = userService.getCurrentUser().getEmail();
       logoutUrl = userService.createLogoutURL(urlToRedirect);
       showNewTab = true;
+      user = new User(userId, userEmail);
     } else {
       loginUrl = userService.createLoginURL(urlToRedirect);
+      user = new User(userId, userEmail);
```
Comment (Froggerson) [inline src/main/java/com/google/sps/servlets/LoginServlet.java]: You can probably put "user = new User(userId, userEmail);" outside of the if statement because the statement is used either way, but honestly it's up to you if you feel like doing that or not.

**Review by Froggerson — APPROVED** [review]
Seems fine to me

File: src/test/java/com/google/sps/UserFoldersServletTest.java (hunk 2 above)
Comment (ethyus) [reply src/test/java/com/google/sps/UserFoldersServletTest.java]: is the function name not clear enough to know what is occurring?

File: src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 3 above)
Comment (ethyus) [reply src/main/java/com/google/sps/servlets/LoginServlet.java]: The value of userId changes based on whether or not the user is logged in.

File: src/test/java/com/google/sps/UserFoldersServletTest.java (hunk 2 above)
Comment (ESaracay) [reply src/test/java/com/google/sps/UserFoldersServletTest.java]: Yeah, I'm not sure if this would be extra but what if you added whats going and what you expect. But this is kind of nit picky I guess.

**Review by ESaracay — APPROVED** [review]

File: src/test/java/com/google/sps/UserFoldersServletTest.java (hunk 2 above)
Comment (ethyus) [reply src/test/java/com/google/sps/UserFoldersServletTest.java]: Made a commit for Change directory structure

File: src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 3 above)
Comment (tylerhou) [reply src/main/java/com/google/sps/servlets/LoginServlet.java]: Agreed with Jennifer here; moving the statement outside is equivalent to the code as you have written currently.

File: src/test/java/com/google/sps/servlets/LoginServletTest.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        .setDefaultHighRepJobPolicyUnappliedJobPercentage(0),
+      new LocalUserServiceTestConfig())
+      .setEnvIsAdmin(true).setEnvIsLoggedIn(true)
+      .setEnvEmail("test@gmail.com").setEnvAuthDomain("gmail.com")
+      .setEnvAttributes(
+        new HashMap(
+          ImmutableMap.of(
+            "com.google.appengine.api.users.UserService.user_id_key", USER_ID)));
+
+  private static final User USER_A = new User(USER_ID, "test@gmail.com");
```
Comment (tylerhou) [inline src/test/java/com/google/sps/servlets/LoginServletTest.java]: LOGGED_IN_USER

File: src/test/java/com/google/sps/servlets/LoginServletTest.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+  private static final User USER_B = new User("null", "null");
```
Comment (tylerhou) [inline src/test/java/com/google/sps/servlets/LoginServletTest.java]: LOGGED_OUT_USER

File: src/test/java/com/google/sps/servlets/LoginServletTest.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+  @After
+  public void tearDown() throws Exception {
+    helper.setEnvIsLoggedIn(true);
+    helper.tearDown();
+  }
+  
+  @Test
+  public void loginWhenUserInDatastoreAlready() throws Exception {
+
```
Comment (tylerhou) [inline src/test/java/com/google/sps/servlets/LoginServletTest.java]: no newline

File: src/test/java/com/google/sps/servlets/LoginServletTest.java [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+  @After
+  public void tearDown() throws Exception {
+    helper.setEnvIsLoggedIn(true);
+    helper.tearDown();
+  }
+  
+  @Test
+  public void loginWhenUserInDatastoreAlready() throws Exception {
+    
+    storeUsertoDatastore(datastore, USER_A);
```
Comment (tylerhou) [inline src/test/java/com/google/sps/servlets/LoginServletTest.java]: naming

File: src/test/java/com/google/sps/servlets/LoginServletTest.java [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+    servlet.doGet(mockRequest, mockResponse);
+    String response = responseWriter.toString();
+    
+    Key key = USER_A.createEntity().getKey(); 
+    USER_A.setUserKey(KeyFactory.keyToString(key));
+    String expectedResponse = new Gson().toJson(getExpectedJsonInfo(USER_A, true, LOGOUTURL, "null"));
```
Comment (tylerhou) [inline src/test/java/com/google/sps/servlets/LoginServletTest.java]:
```java
getExpectedJsonInfo(/*user=*/USER_A, /*tabStatus=*/true, /*logoutUrl=*/LOGOUTURL, /*loginUrl=*/"null")
```

File: src/test/java/com/google/sps/servlets/UserFoldersServletTest.java [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+
+  private final LocalServiceTestHelper helper = 
+    new LocalServiceTestHelper(
+      new LocalDatastoreServiceTestConfig()
+        .setDefaultHighRepJobPolicyUnappliedJobPercentage(0),
+      new LocalUserServiceTestConfig())
+      .setEnvIsAdmin(true).setEnvIsLoggedIn(true)
+      .setEnvEmail("test@gmail.com").setEnvAuthDomain("gmail.com");
+
+  private static final Folder FOLDER_A = new Folder("FIRSTFOLDER", "en");
```
Comment (tylerhou) [inline src/test/java/com/google/sps/servlets/UserFoldersServletTest.java]: We shouldn't hardcode the value of FOLDER_A and FOLDER_B in this case; just put them into each unit test. I can explain why in tomorrow's meeting.

File: src/test/java/com/google/sps/servlets/UserFoldersServletTest.java [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+  public void tearDown() throws Exception {
+    helper.tearDown();
+  }
+  
+  @Test
+  public void QueryUserFolders() throws Exception {
+    // Returns array of Folders and signals front-end to show Create Folder Option
+    
+    // Generate testing User
+    Entity USER_A = new Entity("User", "testId");
```
Comment (tylerhou) [inline src/test/java/com/google/sps/servlets/UserFoldersServletTest.java]: Do we need `_A` here?

File: src/main/java/com/google/sps/data/Folder.java [hunk 11]
```diff
@@ -44,23 +52,16 @@ public Entity createEntity(Key userKey) {
         // Set owner of folder 
         Entity folder = new Entity("Folder", userKey);
 
+        // Store initial folder entity without properties to generate auto ID for entity
```
Comment (tylerhou) [inline src/main/java/com/google/sps/data/Folder.java]: Now this function isn't just creating an entity — it's saving it as well.

Why does this function need to have [side effects](https://en.wikipedia.org/wiki/Side_effect_(computer_science))? Can't the user save the entity themselves?

File: src/main/java/com/google/sps/data/Folder.java [hunk 12]
```diff
@@ -44,23 +52,16 @@ public Entity createEntity(Key userKey) {
...
+        // This is necessary to obtain a "complete" key
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        datastore.put(folder);
+
         folder.setProperty("folderName", this.folderName);
         folder.setProperty("folderDefaultLanguage", this.folderDefaultLanguage);
         folder.setProperty("folderKey", KeyFactory.keyToString(folder.getKey()));
```
Comment (tylerhou) [inline src/main/java/com/google/sps/data/Folder.java]: Why do we need to store the key twice in the entity?

File: src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 13]
```diff
@@ -34,46 +36,54 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
     }
 
-    User user = new User(userId, loginUrl, logoutUrl, userEmail, showNewTab);
-    
     if (userId != "null") {
-        checkUserInDatastore(user);
+      if (!isUserInDatastore(user)) {
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/LoginServlet.java]: `userId != null && !isUserInDatastore(user)`

File: src/test/java/com/google/sps/servlets/UserFoldersServletTest.java (hunk 10 above)
Comment (tylerhou) [reply src/test/java/com/google/sps/servlets/UserFoldersServletTest.java]: Locals should also not use uppercase naming.

File: src/main/java/com/google/sps/data/Folder.java (hunk 12 above)
Comment (ethyus) [reply src/main/java/com/google/sps/data/Folder.java]: The folder entity is keyed by the parent's key (userKey). I store the folder's "complete key" which can be used to retrieve the folder from datastore as well as using the folder's key to set an ancestor path for the cards that are created within the folder. I think this is also convenient because when I convert the entity to a folder object, we have access to the folder's complete key on the front end, which allow us to pass the folder's complete key as a query parameter if the user was to click on the folder. Does that make sense?

File: src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 3 above)
Comment (ethyus) [reply src/main/java/com/google/sps/servlets/LoginServlet.java]: I see what y'all mean!

File: src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 13 above)
Comment (ethyus) [reply src/main/java/com/google/sps/servlets/LoginServlet.java]: Thanks for the catch!

File: src/main/java/com/google/sps/data/Folder.java (hunk 11 above)
Comment (ethyus) [reply src/main/java/com/google/sps/data/Folder.java]: In order for me to create an entity that has a property "folderKey", I would need to store the entity in datastore to generate an auto key identifier. This allows me to call getKey() later.

File: src/test/java/com/google/sps/servlets/LoginServletTest.java (hunk 8 above)
Comment (ethyus) [reply src/test/java/com/google/sps/servlets/LoginServletTest.java]: this is really helpful, thanks

File: src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 3 above)
Comment (zmr) [reply src/main/java/com/google/sps/servlets/LoginServlet.java]: Did this result in a change? Looks like user = new User is still repeated in the if/else clause... Seems like it could go right after the clause...

Comment (ethyus) [reply src/main/java/com/google/sps/servlets/LoginServlet.java]: Fixed!

File: src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 14]
```diff
@@ -34,46 +36,52 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+  public boolean isUserInDatastore(User user) {
 
-    Key userKey = KeyFactory.createKey("User", user.getUserId());
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Key userKey = KeyFactory.createKey("User", user.getUserId());
     Entity userEntity;
 
     // Check if user is in datastore
     try {
       userEntity = datastore.get(userKey);
```
Comment (zmr) [inline src/main/java/com/google/sps/servlets/LoginServlet.java]: is userEntity being used anywhere?

**Comment by zmr** [comment]
can you show the output of the unit test to show they all passed?

File: src/main/java/com/google/sps/data/Folder.java (hunk 12 above)
Comment (tylerhou) [reply src/main/java/com/google/sps/data/Folder.java]: Isn't the "complete key" already stored in the entity by default — you can call `getKey()`?

Comment (ethyus) [reply src/main/java/com/google/sps/data/Folder.java]: Unless I am creating an entity where I manually identify with a unique ID, I can call getKey(). Otherwise, if I am just creating an entity e.g, new Entity("Folder", ancestry key) -- this would still be an incomplete key. That's why I'd need datastore to generate an ID for the entity by calling the put(). I think getKey() requires the entity to have a complete key otherwise it'd throw an error of "incomplete key."

I like your idea though. It is much simpler to call entity.getKey() when I am querying for all the entities from datastore and calling the class method, setKey() in the Folder object during the doGet method.

File: src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 14 above)
Comment (ethyus) [reply src/main/java/com/google/sps/servlets/LoginServlet.java]: It's used to check if the userEntity is in the datastore. If not, we create one

File: src/test/java/com/google/sps/servlets/LoginServletTest.java (hunk 7 above)
Comment (tylerhou) [reply src/test/java/com/google/sps/servlets/LoginServletTest.java]: Ping on this naming.

File: src/test/java/com/google/sps/servlets/LoginServletTest.java [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+    assertTrue(compareJson(response, expectedResponse));
+  }
+
+  @Test
+  public void userNotLoggedIn() throws Exception {
+    helper.setEnvIsLoggedIn(false);
+    servlet.doGet(mockRequest, mockResponse);
+    String response = responseWriter.toString();
+    String expectedResponse = new Gson().toJson(getExpectedJsonInfo(
+        /*user*/LOGGED_OUT_USER, /*tabStatus*/false, /*logoutUrl*/"null", /*loginUrl*/LOGINURL));
```
Comment (tylerhou) [inline src/test/java/com/google/sps/servlets/LoginServletTest.java]: `/*user=*/`, etc.

File: src/test/java/com/google/sps/servlets/UserFoldersServletTest.java [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+    when(mockRequest.getParameter("folderName")).thenReturn("Folder1");
+    when(mockRequest.getParameter("folderDefaultLanguage")).thenReturn("en");
+    when(mockRequest.getParameter("userKey")).thenReturn(userKey);
+
+    servlet.doPost(mockRequest, mockResponse);
+    assertEquals(3, datastore.prepare(new Query("Folder").setAncestor(user.getKey())).countEntities(withLimit(10)));
+
+  }
+
+  public Map<String, Object> getExpectedJsonInfo(List<Folder> folders, boolean showCreateFormStatus) {
```
Comment (tylerhou) [inline src/test/java/com/google/sps/servlets/UserFoldersServletTest.java]: Per the testing best practices — I would strongly suggest [inlining](https://en.wikipedia.org/wiki/Inline_expansion) this function into each method.

Comment (tylerhou) [reply src/test/java/com/google/sps/servlets/UserFoldersServletTest.java]: (because right now showCreateFormStatus and userFolders are essentially test-global constants that are not obvious or implied by the name of the `getExpectedJsonInfo` method.)

Comment (tylerhou) [reply src/test/java/com/google/sps/servlets/UserFoldersServletTest.java]: Actually, even better would be to write out the expected JSON explicitly in each test. That way you don't rely on Gson to encode your response properly (what if Gson has a bug? Then this test won't catch it, as both this test and the servet under test use Gson to encode, so this test will continue to succeed.)

File: src/test/java/com/google/sps/servlets/LoginServletTest.java [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+
+    assertTrue(compareJson(response, expectedResponse));
+  }
+  
+  public Map<String, Object> getExpectedJsonInfo(User user, boolean tabStatus, String logoutUrl, String loginUrl) {
+    Map<String, Object> expectedJsonInfo = new HashMap<>();
```
Comment (tylerhou) [inline src/test/java/com/google/sps/servlets/LoginServletTest.java]: Same here as in the other test.

File: src/main/java/com/google/sps/servlets/UserFoldersServlet.java [hunk 18]
```diff
@@ -35,29 +37,32 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
 
-    if (userService.isUserLoggedIn()) {
+    Map<String, Object> jsonInfo = new HashMap<>();
+    jsonInfo.put("showCreateFormStatus", false);
+    List<Folder> userFolders = new ArrayList<>();
 
-      List<Folder> userFolders = new ArrayList<>();
+    if (userService.isUserLoggedIn()) {
 
+      jsonInfo.put("showCreateFormStatus", true);
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/UserFoldersServlet.java]: Pull this out into a separate constant to avoid typos.

File: src/main/java/com/google/sps/servlets/UserFoldersServlet.java [hunk 19]
```diff
@@ -35,29 +37,32 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+          folder.setFolderKey(KeyFactory.keyToString(entity.getKey()));
           userFolders.add(folder);
         }
       }
-      
-      response.setContentType("application/json;");
-      response.getWriter().println(new Gson().toJson(userFolders));
     }
+
+    jsonInfo.put("userFolders", userFolders);
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/UserFoldersServlet.java]: I would normally suggest using a constant here as well, but there's only one instance so not too big of a deal.

In general, if a string literal is used in multiple places and should be kept in sync, you should use a constant. It makes the literal easier to change and avoid bugs because the compiler will catch the typo for you.

File: src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 20]
```diff
@@ -34,46 +36,52 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
-  public void checkUserInDatastore(User user) {
-    // ** NEED TO WRITE TEST FOR BEHAVIORAL TESTING
+  public boolean isUserInDatastore(User user) {
 
-    Key userKey = KeyFactory.createKey("User", user.getUserId());
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Key userKey = KeyFactory.createKey("User", user.getUserId());
     Entity userEntity;
 
     // Check if user is in datastore
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/LoginServlet.java]: No comment needed.

File: src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 21]
```diff
@@ -34,46 +36,52 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
       loginUrl = userService.createLoginURL(urlToRedirect);
     }
 
-    User user = new User(userId, loginUrl, logoutUrl, userEmail, showNewTab);
-    
-    if (userId != "null") {
-        checkUserInDatastore(user);
+    user = new User(userId, userEmail);
+
+    if ((userId != "null") && (!isUserInDatastore(user))) {
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/LoginServlet.java]: We don't need these extra parens — it's fairly clear to most Java programmers that != binds less tightly than && (otherwise common conditionals like `if (a == 1 && b == 2)` would not work).

File: src/test/java/com/google/sps/servlets/UserFoldersServletTest.java (hunk 16 above)
Comment (ethyus) [reply src/test/java/com/google/sps/servlets/UserFoldersServletTest.java]: very good point. thanks

File: src/test/java/com/google/sps/servlets/LoginServletTest.java (hunk 7 above)
Comment (ethyus) [reply src/test/java/com/google/sps/servlets/LoginServletTest.java]: I changed it to "userLoggedInButNotInDatastore" Is that okay?

File: src/main/java/com/google/sps/servlets/UserFoldersServlet.java [hunk 22]
```diff
@@ -24,26 +24,26 @@
...
 import java.util.HashMap;
 
-
-/** **/
 @WebServlet("/userfolders")
 public class UserFoldersServlet extends HttpServlet {
-
-  /**
+  
+  private final String CREATE_FORM_HEADER = "showCreateFormStatus";
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/UserFoldersServlet.java]: Is this a header?

File: src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 14 above)
Comment (zmr) [reply src/main/java/com/google/sps/servlets/LoginServlet.java]: If I am reading the documentation correctly, datastore.get  returns "null" if the object doesn't exist. The datastore exception only happens if something else goes wrong (server timeout, for example). Which documentation are you looking at for tat says this will throw "EntittyNotFoundException"?

Comment (zmr) [reply src/main/java/com/google/sps/servlets/LoginServlet.java]: Never mind. I was looking at different version of the Client library..

File: src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 23]
```diff
@@ -16,74 +16,69 @@
...
     }
 
-    User user = new User(userId, loginUrl, logoutUrl, userEmail, showNewTab);
-    
-    if (userId != "null") {
-        checkUserInDatastore(user);
+    user = new User(userId, userEmail);
+
+    if (userId != "null" && !isUserInDatastore(user)) {
+      storeUserToDatastore(user);
```
Comment (zmr) [inline src/main/java/com/google/sps/servlets/LoginServlet.java:48]: are there circumstances when the user exists and is logged in but they are not in the datastore?

File: src/main/java/com/google/sps/servlets/UserFoldersServlet.java [hunk 24]
```diff
@@ -21,65 +24,76 @@
...
+          Folder folder = new Folder(entity);
+          folder.setFolderKey(KeyFactory.keyToString(entity.getKey()));
           userFolders.add(folder);
         }
       }
-      
-      response.setContentType("application/json;");
-      response.getWriter().println(new Gson().toJson(userFolders));
+		
+		jsonInfo.put(CREATE_FORM_HEADER, true);
```
Comment (zmr) [inline src/main/java/com/google/sps/servlets/UserFoldersServlet.java]: does this need to be indented this much?

File: src/main/java/com/google/sps/servlets/UserFoldersServlet.java [hunk 25]
```diff
@@ -21,65 +24,76 @@
...
+			DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
       
       String folderName = request.getParameter("folderName");
       String folderDefaultLanguage = request.getParameter("folderDefaultLanguage");
-      String userKey = request.getParameter("userKey");
+
+      String userEmail = userService.getCurrentUser().getEmail();
+      Query userQuery = new Query("User").setFilter(new FilterPredicate("email", FilterOperator.EQUAL, userEmail));
+			Entity userEntity = datastore.prepare(userQuery).asSingleEntity();
+      String userKey = KeyFactory.keyToString(userEntity.getKey());
```
Comment (zmr) [inline src/main/java/com/google/sps/servlets/UserFoldersServlet.java]: this section is the same as what starts on line 50. Rather than repeat the same code, better to separate out in a helper function

File: src/main/java/com/google/sps/servlets/UserFoldersServlet.java (hunk 24 above)
Comment (ethyus) [reply src/main/java/com/google/sps/servlets/UserFoldersServlet.java]: Uhmm that's weird, it seems fine on me IDE... (it was indented correctly)

File: src/main/java/com/google/sps/servlets/UserFoldersServlet.java (hunk 25 above)
Comment (ethyus) [reply src/main/java/com/google/sps/servlets/UserFoldersServlet.java]: Thanks, I just realized that there is this weird bug. When I have my file configured to two spaces for the tab button, the file has weird spacing when it is pushed. This explains why some lines are indented a lot. I'll fix this

File: src/test/java/com/google/sps/servlets/UserFoldersServletTest.java [hunk 26]
```diff
@@ -0,0 +1,180 @@
...
+  }
+  
+  @Test
+  public void queryUserFolders() throws Exception {
+    // Generate testing folders to store in datastore
+    Folder folderA = new Folder(FOLDER_A, FOLDER_LANGUAGE);
+    Folder folderB = new Folder(FOLDER_B, FOLDER_LANGUAGE);
+    
+    // Generate testing user entity to query by email
+	  // And generate a userKey to set as the parent of the testing folders
```
Comment (ethyus) [inline src/test/java/com/google/sps/servlets/UserFoldersServletTest.java]: I don't know why there is spacing on github. On my IDE, it seems fine...

File: src/test/java/com/google/sps/servlets/UserFoldersServletTest.java [hunk 27]
```diff
@@ -0,0 +1,180 @@
...
+          +  "{\"folderName\":\"Second Folder\",\"folderDefaultLanguage\":\"en\",\"folderKey\":\""+ folderBInDatastore.getFolderKey() +"\"}],"
+          + "\"showCreateFormStatus\":true}";
+
+    assertTrue(compareJson(response, expectedResponse));
+  }
+
+  @Test
+  public void userHasNoCurrentFolder() throws Exception {
+    // Generate testing user entity to query by email
+	  // And generate a userKey to set as the parent of the testing folders
```
Comment (ethyus) [inline src/test/java/com/google/sps/servlets/UserFoldersServletTest.java]: Same for this one...

File: src/main/java/com/google/sps/data/Folder.java [hunk 28]
```diff
@@ -1,66 +1,60 @@
...
-    }
+  private String folderName;
+  private String folderDefaultLanguage;
+  private String folderKey;
+  private String parentKey;
+
+  public Folder(
+    String folderName,
+    String folderDefaultLanguage) {
+
```
Comment (tylerhou) [inline src/main/java/com/google/sps/data/Folder.java]: No linebreak.

File: src/main/java/com/google/sps/data/Folder.java [hunk 29]
```diff
@@ -1,66 +1,60 @@
...
+  public void setFolderName(String newFolderName) {
+    this.folderName = newFolderName;
+  }
+
+  public void setFolderKey(String folderKey) {
+    this.folderKey = folderKey;
+  }
+
+  public void setParentKey(String key) {
+    parentKey = key;
```
Comment (tylerhou) [inline src/main/java/com/google/sps/data/Folder.java]: `this`?

File: src/main/java/com/google/sps/servlets/UserFoldersServlet.java [hunk 30]
```diff
@@ -21,65 +22,77 @@
...
+import java.util.HashMap;
 
-
-/** **/
 @WebServlet("/userfolders")
 public class UserFoldersServlet extends HttpServlet {
-
-  /**
+  
+  private static final String CREATE_FORM_HEADER = "showCreateFormStatus";
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/UserFoldersServlet.java]: These aren't really headers; they are keys.

**Review by tylerhou — CHANGES_REQUESTED** [review]

File: src/test/java/com/google/sps/servlets/UserFoldersServletTest.java (hunk 26 above)
Comment (tylerhou) [reply src/test/java/com/google/sps/servlets/UserFoldersServletTest.java]: It's because there is a tab instead of a space. You have your tab-size on your IDE set to 2 spaces; on GitHub they display it as 4.

File: src/main/java/com/google/sps/servlets/UserFoldersServlet.java [hunk 31]
```diff
@@ -21,65 +22,77 @@
...
       
-      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
       datastore.put(folderEntity);
     }
+  }
+
+  public String getUserKey(UserService userService, DatastoreService datastore) {
+    String userEmail = userService.getCurrentUser().getEmail();
+    Query userQuery = new Query("User").setFilter(new FilterPredicate("email", FilterOperator.EQUAL, userEmail));
+	Entity userEntity = datastore.prepare(userQuery).asSingleEntity();
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/UserFoldersServlet.java]: Formatting.

File: src/test/java/com/google/sps/servlets/LoginServletTest.java [hunk 32]
```diff
@@ -0,0 +1,108 @@
...
+public final class LoginServletTest {
+
+  private static final String USER_ID = "testID";
+  private static final User LOGGED_IN_USER = new User(USER_ID, "test@gmail.com");
+  private static final User LOGGED_OUT_USER = new User("null", "null");
+
+  private final LocalServiceTestHelper helper =
+    	new LocalServiceTestHelper(
+      	new LocalDatastoreServiceTestConfig()
+        	.setDefaultHighRepJobPolicyUnappliedJobPercentage(0),
```
Comment (tylerhou) [inline src/test/java/com/google/sps/servlets/LoginServletTest.java]: Has this been run through a formatter?

File: src/test/java/com/google/sps/servlets/LoginServletTest.java [hunk 33]
```diff
@@ -0,0 +1,108 @@
...
+  public void userNotLoggedIn() throws Exception {
+    helper.setEnvIsLoggedIn(false);
+    servlet.doGet(mockRequest, mockResponse);
+    String response = responseWriter.toString();
+    String expectedResponse = "{\"logoutUrl\":\"null\",\"loginUrl\":\"/_ah/login?continue\\u003d%2F\",\"showTabStatus\":false}";
+
+    assertTrue(compareJson(response, expectedResponse));
+  }
+
+  public void storeUsertoDatastore(DatastoreService datastore, User user) {
```
Comment (tylerhou) [inline src/test/java/com/google/sps/servlets/LoginServletTest.java]: Naming.

File: src/test/java/com/google/sps/servlets/UserFoldersServletTest.java [hunk 34]
```diff
@@ -0,0 +1,180 @@
...
+import java.util.ArrayList;
+
+@RunWith(JUnit4.class)
+public final class UserFoldersServletTest {
+
+  private static final String USER_KIND = "User";
+  private static final String USER_ID = "testId";
+  private static final String FOLDER_A = "First Folder";
+  private static final String FOLDER_B = "Second Folder";
+  private static final String FOLDER_LANGUAGE = "en";
```
Comment (tylerhou) [inline src/test/java/com/google/sps/servlets/UserFoldersServletTest.java]: These are small enough that they should be hardcoded into each test.

**Review by tylerhou — CHANGES_REQUESTED** [review]

File: src/test/java/com/google/sps/servlets/UserFoldersServletTest.java (hunk 34 above)
Comment (tylerhou) [reply src/test/java/com/google/sps/servlets/UserFoldersServletTest.java]: Also, their names don't give the test reader enough context to understand what they contain, so they also don't follow the rules I posted in the chat about testing constants.

File: src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 35]
```diff
@@ -16,74 +16,69 @@
...
     String urlToRedirect = "/";
     String userId = "null";
     String logoutUrl = "null";
     String loginUrl = "null";
     String userEmail = "null";
-
-    // Boolean to update User's main page 
     Boolean showNewTab = false;
 
+    User user;
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/LoginServlet.java]: Declare this on line 46.

File: src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 36]
```diff
@@ -16,74 +16,69 @@
...
+  public boolean isUserInDatastore(User user) {
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    Entity userEntity;
+    Key userKey = KeyFactory.createKey("User", user.getUserId());
 
-    // Check if user is in datastore
     try {
-      userEntity = datastore.get(userKey);
+      datastore.get(userKey);
       user.setUserKey(KeyFactory.keyToString(userKey));
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/LoginServlet.java:66]: You don't have to do this in this PR, but I would suggest restructuring this method. Currently, the name of the method suggests that it's a pure function — it looks up whether the user is in Datastore. But it actually has side effects — it alters user itself.

I think a better way to write this function to eliminate side effects is to return the Key if the user exists in Datastore — otherwise, return null. Then the calling code can use this to determine whether the user already exists, and the side effect can be done there too.

File: src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 37]
```diff
@@ -16,74 +16,69 @@
...
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
-    // Create new user entity 
     Entity newEntity = user.createEntity();
-
-    // Create User's key and add to object
-    Key key = newEntity.getKey(); 
+    Key key = newEntity.getKey();
     String userKeyStr = KeyFactory.keyToString(key);
     user.setUserKey(userKeyStr);
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/LoginServlet.java:79]: Same here as above. Again, doesn't have to be this PR, but this method should return the key and let the calling code change the user.

Keeping methods pure has two benefits.

1) The method is more reusable, because sometimes you don't want/need those side effects.
2) It's easier to reason about pure functions, because you don't have to be worried that they are changing state.

File: src/main/java/com/google/sps/servlets/UserFoldersServlet.java [hunk 38]
```diff
@@ -19,67 +20,80 @@
...
-      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
       PreparedQuery results = datastore.prepare(folderQuery);
 
-      // Add folders to a list
       if (results != null) {
-        for (Entity entity: results.asIterable()) {
-          Folder folder = Folder.EntityToFolder(entity);
+        for (Entity entity : results.asIterable()) {
+          Folder folder = new Folder(entity);
+          folder.setFolderKey(KeyFactory.keyToString(entity.getKey()));
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/UserFoldersServlet.java:61]: Why is this line here? Why can't this be done in the constructor?

**Review by tylerhou — APPROVED** [review]

File: src/main/java/com/google/sps/servlets/UserFoldersServlet.java (hunk 38 above)
Comment (ethyus) [reply src/main/java/com/google/sps/servlets/UserFoldersServlet.java:61]: I restructured this in tdd5

File: src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 36 above)
Comment (ethyus) [reply src/main/java/com/google/sps/servlets/LoginServlet.java:66]: Thank you! I'll fix this in a different branch. I'll maintain this concept in the future
